### PR TITLE
Enhance AKS external cluster status information

### LINKED ISSUE #10853: AKS external cluster: enhance Node Pool status information
**_First Issue:_**
Sometimes, Azure node pool information is not precise when editing a node pool. For instance, when I increased the number of replicas from 1 to 2, it worked after some time, but "Provisioning state" was stating as "Failed" in Azure console, which was reflecting in Kubermatic.
image (1)
image

The same can happen when creating a new node pool, or editing the Kubelet version for a running node pool.

**Suggestion**: add extra information in the Kubermatic UI regarding both Azure's statuses (power state and provisioning state)

**_Second Issue:_**
Sometimes, when you try to edit a node pool (increase the number of replicas for a MD or change the Kubelet version), it keeps loading for a long time, then come back to previous configuration, without any feedback (sometimes only showing "error" status).

**Suggestion**: fetch error events somehow for the node pool and display them in the UI.

### PR DESCRIPTION
**Pull request by hbernardo**
**What this PR does / why we need it**:
It's confusing for the user when something went wrong on Azure side, because we show only provisioning state without any message. This PR enhances the feedback on the API side regarding AKS external cluster status information on errors or warning states.

**Which issue(s) this PR fixes**:

Fixes #10853

**What type of PR is this?**

/kind feature
/kind api-change

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
Enhanced AKS external cluster status information on errors or warning states.
```

**Documentation**:

```documentation
NONE
```

### CODE REVIEW

File: pkg/provider/cloud/aks/provider.go [hunk 1]
```diff
@@ -225,7 +237,7 @@ func ConvertStatus(provisioningState string, powerState armcontainerservice.Code
 	case provisioningState == string(resources.SucceededAKSState) && powerState == armcontainerservice.Code(resources.StoppedAKSState):
 		return apiv2.StoppedExternalClusterState
 	case provisioningState == string(resources.FailedAKSState):
```
Comment (imharshita) [inline pkg/provider/cloud/aks/provider.go]: lets remove this condition for error and replace it with warning:
`case provisioningState == string(resources.FailedAKSState) && string(powerState) == string(resources.RunningAKSMDState)`
same for md.
idea behind this this if cluster or md power state is running then we should not show overall state of the object as `Failed`

File: pkg/handler/v2/external_cluster/aks.go [hunk 2]
```diff
@@ -651,9 +651,21 @@ func createMachineDeploymentFromAKSNodePoll(nodePool *armcontainerservice.Manage
...
+	if nodePool.ProvisioningState != nil && (nodePool.PowerState != nil && nodePool.PowerState.Code != nil) {
 		md.Phase = apiv2.ExternalClusterMDPhase{
 			State: aks.ConvertMDStatus(*nodePool.ProvisioningState, *nodePool.PowerState.Code),
+			AKS: &apiv2.AKSMDPhase{
+				ProvisioningState: apiv2.AKSProvisioningState(*nodePool.ProvisioningState),
+				PowerState:        apiv2.AKSPowerState(*nodePool.PowerState.Code),
+			},
+		}
+		switch {
+		case string(*nodePool.PowerState.Code) == string(resources.StoppedAKSMDState) && *nodePool.ProvisioningState == string(resources.SucceededAKSMDState):
```
Comment (imharshita) [inline pkg/handler/v2/external_cluster/aks.go]: message could be something like this:
`The Kubernetes service is currently stopped.
You can only start or delete a stopped AKS cluster. To perform any operation like scale or upgrade, start your cluster first.`

File: pkg/handler/v2/external_cluster/aks.go [hunk 3]
```diff
@@ -651,9 +651,21 @@ func createMachineDeploymentFromAKSNodePoll(nodePool *armcontainerservice.Manage
...
+			md.Phase.StatusMessage = "Power state is marked as 'Stopped' by Azure."
+		case string(*nodePool.PowerState.Code) == string(resources.StoppedAKSMDState) && *nodePool.ProvisioningState == string(resources.FailedAKSMDState):
+			md.Phase.StatusMessage = "Power state is marked as 'Stopped' and provisioning state is marked as 'Failed'. Please check on Azure side for previous operations on the node pool to resolve any failures."
+		case string(*nodePool.PowerState.Code) == string(resources.RunningAKSMDState) && *nodePool.ProvisioningState == string(resources.FailedAKSMDState):
+			md.Phase.StatusMessage = "Despite the power state is 'Running', provisioning state is marked as 'Failed'. Please check on Azure side for previous operations on the node pool to resolve any failures."
 		}
 	}
```
Comment (imharshita) [inline pkg/handler/v2/external_cluster/aks.go]:
```suggestion
                state := aks.ConvertMDStatus(*nodePool.ProvisioningState, *nodePool.PowerState.Code)
		md.Phase = apiv2.ExternalClusterMDPhase{
			State: state,
			AKS: &apiv2.AKSMDPhase{
				ProvisioningState: apiv2.AKSProvisioningState(*nodePool.ProvisioningState),
				PowerState:        apiv2.AKSPowerState(*nodePool.PowerState.Code),
			},
		}
		switch {
		case state == apiv2.StoppedExternalClusterState:
			md.Phase.StatusMessage = "The Kubernetes service is currently stopped. You can only start or delete a stopped AKS cluster. To perform any operation like scale or upgrade, start your cluster first."
		case state == apiv2.WarningExternalClusterState:
			md.Phase.StatusMessage = "The last operation attempted on this cluster failed. The nodes may still be running. Check previous operations on the cluster to resolve any failures."
		}
	}
```

Comment (imharshita) [reply pkg/handler/v2/external_cluster/aks.go]: plz make same changes to md, by replacing `cluster = > node pool` in statusMessage

**Comment by imharshita** [comment]
added a few comments, still reviewing..

File: pkg/provider/cloud/aks/provider.go [hunk 4]
```diff
@@ -225,7 +237,7 @@ func ConvertStatus(provisioningState string, powerState armcontainerservice.Code
...
-		return apiv2.ErrorExternalClusterState
+		return apiv2.ErrorExternalClusterState // TODO: think about changing this to "Warning" and add other variables to decide if actual error, because there are cases that Azure states "Failed" for Provision State but actually it's working/usable
```
Comment (imharshita) [inline pkg/provider/cloud/aks/provider.go]:
```suggestion
		case provisioningState == string(resources.StartingAKSState):
		return apiv2.StartingExternalClusterState
```
lets add `apiv2.StartingExternalClusterState` for both cluster and md

File: pkg/api/v2/types.go [hunk 5]
```diff
@@ -777,6 +801,7 @@ type ExternalClusterMachineDeployment struct {
 type ExternalClusterMDPhase struct {
```
Comment (imharshita) [inline pkg/api/v2/types.go:807]: should we rename this ExternalClusterMDPhase => ExternalClusterMDStatus. wdyt? 
If yes, then we can rename 
AKSMDPhase  => AKSMDStatus

File: pkg/provider/cloud/aks/provider.go [hunk 6]
```diff
@@ -235,14 +245,18 @@ func ConvertStatus(provisioningState string, powerState armcontainerservice.Code
...
+func ConvertMDStatus(provisioningState string, powerState armcontainerservice.Code, allNodesRunning bool) apiv2.ExternalClusterMDState {
 	switch {
 	case provisioningState == string(resources.CreatingAKSMDState):
 		return apiv2.ProvisioningExternalClusterMDState
 	case provisioningState == string(resources.SucceededAKSMDState) && string(powerState) == string(resources.RunningAKSMDState):
 		return apiv2.RunningExternalClusterMDState
-	case provisioningState == string(resources.FailedAKSMDState):
+	case provisioningState == string(resources.SucceededAKSMDState) && string(powerState) == string(resources.StoppedAKSMDState):
+		return apiv2.StoppedExternalClusterMDState
+	case provisioningState == string(resources.FailedAKSMDState) && !allNodesRunning:
```
Comment (imharshita) [inline pkg/provider/cloud/aks/provider.go]: if will think about this condition:
this basically says, if last operation failed and readyreplicas != replicas, then md state = failed,
suppose if last Upgrade operation fails, and hence readyreplicas != replicas , but still some nodes are ready, should we declare MD as Failed/Unusable

maybe when provisioningState == string(resources.FailedAKSMDState) and ready replicas == 0, then we can declare md as unusable
wdyt?

**Review by imharshita — CHANGES_REQUESTED** [review]

File: pkg/provider/cloud/aks/provider.go (hunk 4 above)
Comment (hbernardo) [reply pkg/provider/cloud/aks/provider.go]: This case is already being handled by line 231:
```
	case provisioningState == string(resources.StartingAKSState):
		return apiv2.ProvisioningExternalClusterState
```

Comment (hbernardo) [reply pkg/provider/cloud/aks/provider.go]: Anyway, I decided to add a similar one for MD

**Pull request merged**